architecture/2024-08-applications-rp-component-threat-model.md

```diff
@@ -0,0 +1,224 @@
+| RP                 | Resource Provider    |
+| UCP                 | Universal Control Plane for Radius   |
+
+## System Description
+
+Applications RP is a Radius component that acts as resource provider for application and its resources. It communicates over HTTP. The RP has a Datastore for storing Radius data, Message Queue for processing asynchronous request and a Secret Store for storing sensitive information such as certficates. All these are configurable components and support multiple implementations. Users and Clients cannot directly communicate with Applications RP. They instead communicate with UCP. UCP forwards relevant requests to Applications RP. Applications may have Kubernetes resources. Applications RP manages these resources on the user's behalf. This may launch user application's code on the same cluster as Radius, or a different cluster. It also has access to user's cloud credentials and manages user's cloud resources. Applications RP can invoke *recipes* which are bicep or terraform code. These recipes are used to deploy application infrastructure components like databases.  
+
+### Architecture
+
+Application RP recieves HTTP requests from UCP. It does not interface directly with user/ cli. These requests have untrusted json payloads. RP validates these payloads before consuming them. 
+
+The RP consists of four types of resource providers for managing various types of resources in an application. `Applications.Core` resource provider manages core application resources such as application, environment, container and gateways. `Applications.Dapr` resource provider manages all dapr resources that are deployed as part of application. `Applications.Datastore` resource provider supports provisioning SQL database, Mongo DB and Redis Cache.
+`Applications.Messaging` resources provider manages queues such as Rabbit MQ.
+
+Applications RP has a key sub component `Recipe Engine` to execute `recipes`. 
+`Recipes` are Bicep or Terraform code supplied by user that is used to deploy infrastructure components on Azure and AWS. The Bicep recipes are fetched from OCI compliant registries. Terraform recipes are public modules and fetched from internet too. 
+
+In order to execute Terraform recipes, Applications RP installs latest Terraform. It also mounts an empty directory `/terraform` into Applications RP pod. It uses this directory for executing terraform recipes using the installed executable. The output resources generated from terraform module are converted to Radius output resources and stored in our datastore. 
+
+In order to deploy bicep recipes, Applications RP sends a request to UCP, which in turn forwards it to Deployment Engine. 
+
+The RP uses a queue to process requests asyncronously. Information about resources that are deployed / being deployed is stored in a datastore. 
+
```

Comment:
Let's talk about **how** we communicate and manage things in the cloud. 

- Recipes
- Things we managed directly (volumes, keyvault, managed identities)

Let's talk about **how** we communicate and manage things in Kubernetes. 

Example:

When a users deploys a gateway resource. Application RP uses its Kubernetes `ServiceAccount` to authenticate with the local Kubernetes cluster, and deploys `HTTPProxy` (Contour) resources to open an listener to the internet, and route traffic to the user's application. 


Example:

When a user deploys a container resource that needs a managed identity.... (please go find out)